File: Python/백준/Silver/6588. 골드바흐의 추측/골드바흐의 추측.py
import sys
from math import sqrt

def a():
  all_prime = []
  prime_table = [True for _ in range(1000001)]
  prime_table[0] = False
  prime_table[1] = False
  for i in range(2, 1000001):
    if prime_table[i]:
      all_prime.append(i)
      for j in range(2, 1000001):
        if i * j <= 1000000:
          prime_table[i * j] = False
        else: break
  
  while True:
    N = int(sys.stdin.readline())
    if N == 0: break
    a, b = -1, -1
    for i in all_prime:
      if i < 500001:
        if prime_table[N - i]:
          a, b = i, N - i
          break
      else: 
        break
    if a == -1:
      print("Goldbach's conjecture is wrong.")
    else:
      print(N, "=", a, "+", b)
a()

File: Java/백준/Silver/6588. 골드바흐의 추측/골드바흐의 추측.java
import java.util.*;
import java.io.*;
import java.lang.*;

class Solution {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static Map<Integer, Boolean> hm = new HashMap<>();
    public void solution() throws IOException{
        while (true) {
            int N = Integer.parseInt(br.readLine());
            if (N == 0) break;
            boolean chk = false;
            int a = 0;
            int b = 0;
            for (int i = 3; i <= N / 2; i += 2) {
                a = i;
                b = N - i;
                chk = isPrime(b) && isPrime(a);
                if (chk) break;
            }
            if (chk) bw.write(N + " = " + a + " + " + b + "\n");
            else bw.write("Goldbach's conjecture is wrong.\n");
        } 
        bw.flush();
        bw.close();
    }

    
    public boolean isPrime(int n) {
        if (hm.containsKey(n)) return hm.get(n);
        if (n == 1) return false;
        else if (n == 2) return true;
        int sqrt = (int)Math.sqrt(n) + 1;
        for (int i = 2; i <= sqrt; i++) {
            if (n % i == 0) {
                hm.put(n, false);
                return false;
            }
        }
        hm.put(n, true);
        return true;
    }
}

public class Main {
    public static void main(String argv[]) throws IOException{
        Solution sol = new Solution();
        sol.solution();
    }
}